fix data race in vvm/TestBasic/LeadershipCollision

https://github.com/voedger/voedger/actions/runs/13918370733/job/38945679053?pr=3446
```
WARNING: DATA RACE
  Read at 0x00c0007fe090 by goroutine 2970:
    runtime.mapaccess2_faststr()
        /opt/hostedtoolcache/go/1.23.7/x64/src/runtime/map_faststr.go:117 +0x0
    github.com/voedger/voedger/pkg/istorage/mem.(*appStorage).readPart()
        /home/runner/work/voedger/voedger/pkg/istorage/mem/impl.go:288 +0x166
    github.com/voedger/voedger/pkg/istorage/mem.(*appStorage).Read()
        /home/runner/work/voedger/voedger/pkg/istorage/mem/impl.go:316 +0x144
    github.com/voedger/voedger/pkg/istoragecache.(*cachedAppStorage).Read()
        /home/runner/work/voedger/voedger/pkg/istoragecache/impl.go:365 +0x224
    github.com/voedger/voedger/pkg/istructsmem/internal/vers.(*Versions).Prepare()
        /home/runner/work/voedger/voedger/pkg/istructsmem/internal/vers/impl.go:41 +0x1c6
    github.com/voedger/voedger/pkg/istructsmem.(*AppConfigType).prepare()
        /home/runner/work/voedger/voedger/pkg/istructsmem/appstruct-types.go:147 +0x352
    github.com/voedger/voedger/pkg/istructsmem.(*appStructsProviderType).BuiltIn()
        /home/runner/work/voedger/voedger/pkg/istructsmem/impl.go:61 +0x34c
    github.com/voedger/voedger/pkg/vvm.VVMAppsBuilder.BuildAppsArtefacts()
        /home/runner/work/voedger/voedger/pkg/vvm/impl_builder.go:76 +0x503
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:30 +0x372
    github.com/alecthomas/participle/v2.isLeftRecursive.func2()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:52 +0x19b
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.isLeftRecursive()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:33 +0x169
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:17 +0xb4
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:40 +0x284
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:26 +0x165
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:13 +0x5bd
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:26 +0x165
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:48 +0xba
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:26 +0x165
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:30 +0x372
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:26 +0x165
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:34 +0x3c8
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:26 +0x165
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:34 +0x3c8
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:26 +0x165
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:19 +0x[214](https://github.com/voedger/voedger/actions/runs/13918370733/job/38945679053?pr=3446#step:8:216)
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:26 +0x165
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:34 +0x3c8
    github.com/alecthomas/participle/v2.isLeftRecursive.func2()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:52 +0x19b
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.isLeftRecursive()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:33 +0x169
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:17 +0xb4
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:40 +0x284
    github.com/alecthomas/participle/v2.validate.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:26 +0x165
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:13 +0x5bd
    github.com/alecthomas/participle/v2.isLeftRecursive.func2()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/validate.go:52 +0x19b
    github.com/alecthomas/participle/v2.visit()
        /home/runner/go/pkg/mod/github.com/alecthomas/participle/v2@v2.1.1/visit.go:9 +0xec
    github.com/alecthomas/participle/v2.visit.func1()
        /home/runner/work/voedger/voedger/pkg/vvm/wire_gen.go:143 +0x14ec
    github.com/voedger/voedger/pkg/vvm.Provide()
        /home/runner/work/voedger/voedger/pkg/vvm/wire_gen.go:299 +0xbc7
    github.com/voedger/voedger/pkg/vvm.TestBasic.func2()
        /home/runner/work/voedger/voedger/pkg/vvm/impl_orch_test.go:55 +0x35a
    testing.tRunner()
        /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:1690 +0x226
    testing.(*T).Run.gowrap1()
        /opt/hostedtoolcache/go/1.23.7/x64/src/testing/testing.go:1743 +0x44

  Previous write at 0x00c0007fe090 by goroutine 2247:
    runtime.mapassign_faststr()
        /opt/hostedtoolcache/go/1.23.7/x64/src/runtime/map_faststr.go:223 +0x0
    github.com/voedger/voedger/pkg/istorage/mem.(*appStorage).Put()
        /home/runner/work/voedger/voedger/pkg/istorage/mem/impl.go:219 +0x1e8
    github.com/voedger/voedger/pkg/istorage/mem.(*appStorage).PutBatch()
        /home/runner/work/voedger/voedger/pkg/istorage/mem/impl.go:242 +0x2d0
    github.com/voedger/voedger/pkg/istoragecache.(*cachedAppStorage).PutBatch()
        /home/runner/work/voedger/voedger/pkg/istoragecache/impl.go:269 +0x1e6
    github.com/voedger/voedger/pkg/istructsmem.(*appViewRecords).PutBatch()
        /home/runner/work/voedger/voedger/pkg/istructsmem/viewrecords-types.go:151 +0x33e
    github.com/voedger/voedger/pkg/sys/storages.(*viewRecordsStorage).ApplyBatch()
        /home/runner/work/voedger/voedger/pkg/sys/storages/impl_view_records_storage.go:137 +0x8a8
    github.com/voedger/voedger/pkg/state/stateprovide.(*bundledHostState).FlushBundles()
        /home/runner/work/voedger/voedger/pkg/state/stateprovide/impl_bundled_host_state.go:138 +0x2d3
    github.com/voedger/voedger/pkg/state/stateprovide.(*asyncActualizerState).FlushBundles()
        <autogenerated>:1 +0x3e
    github.com/voedger/voedger/pkg/processors/actualizers.(*asyncProjector).flush()
        /home/runner/work/voedger/voedger/pkg/processors/actualizers/async.go:559 +0x41b
    github.com/voedger/voedger/pkg/processors/actualizers.(*asyncProjector).Flush()
        /home/runner/work/voedger/voedger/pkg/processors/actualizers/async.go:491 +0x244
    github.com/voedger/voedger/pkg/pipeline.p_flush()
        /home/runner/work/voedger/voedger/pkg/pipeline/async.go:65 +0x164
    github.com/voedger/voedger/pkg/pipeline.puller_async()
        /home/runner/work/voedger/voedger/pkg/pipeline/async.go:51 +0x1fe
    github.com/voedger/voedger/pkg/pipeline.NewAsyncPipeline.gowrap1()
        /home/runner/work/voedger/voedger/pkg/pipeline/async-pipeline-impl.go:54 +0x33
  
  
  Goroutine [224](https://github.com/voedger/voedger/actions/runs/13918370733/job/38945679053?pr=3446#step:8:226)7 (running) created at:
    github.com/voedger/voedger/pkg/pipeline.NewAsyncPipeline()
        /home/runner/work/voedger/voedger/pkg/pipeline/async-pipeline-impl.go:54 +0xade
    github.com/voedger/voedger/pkg/processors/actualizers.(*asyncActualizer).init()
        /home/runner/work/voedger/voedger/pkg/processors/actualizers/async.go:231 +0x2399
    github.com/voedger/voedger/pkg/processors/actualizers.(*asyncActualizer).Run()
        /home/runner/work/voedger/voedger/pkg/processors/actualizers/async.go:98 +0xa5
    github.com/voedger/voedger/pkg/processors/actualizers.(*actualizers).NewAndRun()
        /home/runner/work/voedger/voedger/pkg/processors/actualizers/actualizers.go:54 +0x4ed
    github.com/voedger/voedger/pkg/appparts.IActualizerRunner.NewAndRun-fm()
        <autogenerated>:1 +0x103
    github.com/voedger/voedger/pkg/appparts/internal/actualizers.(*PartitionActualizers).start.func1()
        /home/runner/work/voedger/voedger/pkg/appparts/internal/actualizers/actualizers.go:97 +0x256
  ==================
  --- FAIL: TestBasic (12.48s)
      --- FAIL: TestBasic/LeadershipCollision (4.79s)
          testing.go:1399: race detected during execution of test
```

--- a/pkg/istorage/mem/impl.go
+++ b/pkg/istorage/mem/impl.go
@@ -15,25 +15,36 @@ import (
 	"github.com/voedger/voedger/pkg/istorage"
 )
 
+// need to share the lock between different appStorage instances for the case when the mem storage is shared between few VVMs
+// otherwise few VVMs will use the same map from appStorageFactory, but RWMutex'es are different per each appStorage instance!
+// see https://github.com/voedger/voedger/issues/3447
+type storageWithLock struct {
+	data map[string]map[string]coreutils.DataWithExpiration
+	lock sync.RWMutex
+}
+
 type appStorageFactory struct {
-	storages map[string]map[string]map[string]coreutils.DataWithExpiration
+	storages map[string]*storageWithLock
 	iTime    coreutils.ITime
 }
 
 func (s *appStorageFactory) AppStorage(appName istorage.SafeAppName) (istorage.IAppStorage, error) {
-	storage, ok := s.storages[appName.String()]
+	storageWithLock, ok := s.storages[appName.String()]
 	if !ok {
 		return nil, istorage.ErrStorageDoesNotExist
 	}
 
-	return &appStorage{storage: storage, iTime: s.iTime}, nil
+	return &appStorage{storage: storageWithLock.data, iTime: s.iTime, lock: &storageWithLock.lock}, nil
 }
 
 func (s *appStorageFactory) Init(appName istorage.SafeAppName) error {
 	if _, ok := s.storages[appName.String()]; ok {
 		return istorage.ErrStorageAlreadyExists
 	}
-	s.storages[appName.String()] = map[string]map[string]coreutils.DataWithExpiration{}
+	s.storages[appName.String()] = &storageWithLock{
+		data: map[string]map[string]coreutils.DataWithExpiration{},
+		lock: sync.RWMutex{},
+	}
 
 	return nil
 }
@@ -46,7 +57,7 @@ func (s *appStorageFactory) StopGoroutines() {}
 
 type appStorage struct {
 	storage      map[string]map[string]coreutils.DataWithExpiration
-	lock         sync.RWMutex
+	lock         *sync.RWMutex
 	testDelayGet time.Duration // used in tests only
 	testDelayPut time.Duration // used in tests only
 	iTime        coreutils.ITime

--- a/pkg/istorage/mem/provide.go
+++ b/pkg/istorage/mem/provide.go
@@ -11,7 +11,7 @@ import (
 
 func Provide(iTime coreutils.ITime) istorage.IAppStorageFactory {
 	return &appStorageFactory{
-		storages: map[string]map[string]map[string]coreutils.DataWithExpiration{},
+		storages: map[string]*storageWithLock{},
 		iTime:    iTime,
 	}
 }

